File: scripts/test_solutions/297_serialize_deserialize_tree.py
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Codec:
    def serialize(self, root):
        result = []

        def dfs(node):
            if not node:
                result.append('null')
            else:
                result.append(str(node.val))
                dfs(node.left)
                dfs(node.right)

        dfs(root)
        return ','.join(result)

    def deserialize(self, data):
        vals = iter(data.split(','))

        def dfs():
            val = next(vals)
            if val == 'null':
                return None
            node = TreeNode(int(val))
            node.left = dfs()
            node.right = dfs()
            return node

        return dfs()

File: scripts/test_solutions/297_serialize_deserialize_tree.js
var serialize = function(root) {
    const result = [];
    function dfs(node) { if (!node) result.push('null'); else { result.push(node.val); dfs(node.left); dfs(node.right); } }
    dfs(root);
    return result.join(',');
};
var deserialize = function(data) {
    const vals = data.split(',');
    let i = 0;
    function dfs() {
        if (vals[i] === 'null') { i++; return null; }
        const node = { val: parseInt(vals[i++]), left: null, right: null };
        node.left = dfs(); node.right = dfs();
        return node;
    }
    return dfs();
};
